nntrainer/tensor/manager.cpp

```diff
@@ -35,37 +61,148 @@ void Manager::trackWeights(std::vector<Weight> &ws) {
   layer_weights.reserve(ws.size());
 
   size_t weight_size = 0;
+  size_t grad_size = 0;
 
   for (auto &w : ws) {
     layer_weights.emplace_back(std::ref(w));
+    size_t len = w.getDim().getDataLen();
+    weight_size += len;
     if (w.getTrainable())
-      weight_size += w.getDim().getDataLen();
+      grad_size += len;
   }
 
   weights.push_back(layer_weights);
 
-  max_weight_size = std::max(max_weight_size, weight_size);
+  total_weight_size += weight_size;
+  total_grad_size += grad_size;
+  max_grad_size = std::max(max_grad_size, grad_size);
 }
 
 /**
  * @brief Allocate and initialize the weight variable
  */
 void Manager::initialize() {
-  Tensor shared_grad;
-  if (max_weight_size > 0 && enable_gradient_memory_opt)
-    shared_grad = Tensor(max_weight_size);
+  if (total_weight_size == 0) {
+    ml_logw("Nothing done on initialize because there is no weight registered");
+    return;
+  }
+  using AllocFunc = std::function<Tensor(const TensorDim &, size_t)>;
+
+  AllocFunc allocate_none = [](const TensorDim &dim, size_t) {
+    return Tensor();
+  };
+
+  AllocFunc allocate_weight = allocate_none;
+  AllocFunc allocate_grad = allocate_none;
+
+  if (use_shared_memory) {
+    size_t grad_size =
+      enable_gradient_memory_opt ? max_grad_size : total_grad_size;
+    size_t total_size = total_weight_size + grad_size;
```

Comment:
how do you handle the scenario when enable_gradient_memory_opt is false and a layer's weights are set to trainable which earlier were not trainable and vice versa.

Reply:
We made a offline discussion to manage two mmap (one for gradient, and another one for weights in case of reallocation happens)

Reply:
Done with a new commit. Please take another look